### PR TITLE
[BE] Public/private job

### PR DESCRIPTION
**Pull request by MihaiDruta**
[BE] Public/private job
https://trello.com/c/0yC8EhYU/42-be-public-private-job

Job could be private/public. Implemented function to invite a user to a job. Users could see only available activities in activity list.

### CODE REVIEW

**Review by nspc97 — APPROVED** [review]

File: src/Entity/Activity.php [hunk 1]
```diff
@@ -300,4 +315,14 @@ public function updatedTimestamps(): void
             $this->setCreatedAt(new DateTime('now'));
         }
     }
+
+    public function isPublic(): bool
+    {
+        return $this->public;
+    }
+
+    public function setPublic($public): void
```
Comment (SpartakusMd) [inline src/Entity/Activity.php]: `setPublic(bool $public)`

File: src/Migrations/Version20190619140837.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        return '';
+    }
+
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql',
+            'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE activity ADD privacy TINYINT(1) NOT NULL');
```
Comment (SpartakusMd) [inline src/Migrations/Version20190619140837.php]: `public` not `privacy`.

Comment (SpartakusMd) [reply src/Migrations/Version20190619140837.php]: Actualy I saw next migration in which you do the renaming. Remove both of them so you have one migration which is correct from the start.

File: src/Security/AccessRightsPolicy.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+class AccessRightsPolicy
+{
+
+    /**
+     * @var ActivityUserRepository
+     */
+    private $repository;
+
+    public function __construct(ActivityUserRepository $repository)
```
Comment (SpartakusMd) [inline src/Security/AccessRightsPolicy.php]: Use full variable name `$activityUserRepository` so later you know for which entity is this repository without scrolling to the constructor.

File: src/Security/AccessRightsPolicy.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    public function checkRightsToActivity(Activity $activity, User $user): bool
+    {
+        $userToActivity = $this->repository->findBy(array('activity' => $activity, 'user' => $user));
+
+        if (empty($userToActivity) && ($activity->isPublic() === false) && $user !== $activity->getOwner()) {
+            return false;
+        }
+        return true;
+    }
+}
```
Comment (SpartakusMd) [inline src/Security/AccessRightsPolicy.php]: Add empty line at the end.

File: src/Security/AccessRightsPolicy.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    private $repository;
+
+    public function __construct(ActivityUserRepository $repository)
+    {
+        $this->repository = $repository;
+    }
+
+    public function checkRightsToActivity(Activity $activity, User $user): bool
+    {
+        $userToActivity = $this->repository->findBy(array('activity' => $activity, 'user' => $user));
```
Comment (SpartakusMd) [inline src/Security/AccessRightsPolicy.php]: Do the check `$activity->isPublic() === false) && $user !== $activity->getOwner()` before checking if user is invited/has applied. Most of the jobs will be public and the request to the database won't be needed.

File: src/DTO/ActivityDTO.php [hunk 6]
```diff
@@ -88,6 +87,22 @@ class ActivityDTO
      */
     public $status = self::STATUS_NEW;
 
+
+    /**
+     * @var boolean
+     * @Serializer\Type("integer")
```
Comment (SpartakusMd) [inline src/DTO/ActivityDTO.php]: Type `boolean` maybe? https://jmsyst.com/libs/serializer/master/reference/annotations#type

**Review by SpartakusMd — CHANGES_REQUESTED** [review]

**Review by nspc97 — APPROVED** [review]

File: src/Security/AccessRightsPolicy.php [hunk 7]
```diff
@@ -14,18 +14,20 @@ class AccessRightsPolicy
...
+        $this->repository = $activityUserRepository;
     }
 
     public function checkRightsToActivity(Activity $activity, User $user): bool
     {
-        $userToActivity = $this->repository->findBy(array('activity' => $activity, 'user' => $user));
 
-        if (empty($userToActivity) && ($activity->isPublic() === false) && $user !== $activity->getOwner()) {
-            return false;
+        if (($activity->isPublic() === false) && $user !== $activity->getOwner()) {
```
Comment (SpartakusMd) [inline src/Security/AccessRightsPolicy.php]: Use early return. You should always avoid nesting if statements.

**Review by SpartakusMd — CHANGES_REQUESTED** [review]

**Review by SpartakusMd — APPROVED** [review]

File: src/DTO/ActivityDTO.php [hunk 8]
```diff
@@ -88,6 +87,22 @@ class ActivityDTO
      */
     public $status = self::STATUS_NEW;
 
+
+    /**
+     * @var boolean
+     * @Serializer\Type("boolean")
+     * @Assert\Range(
```
Comment (SpartakusMd) [inline src/DTO/ActivityDTO.php]: This should not be required and should not work.

File: src/Security/AccessRightsPolicy.php [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+     * @var ActivityUserRepository
+     */
+    private $repository;
+
+    public function __construct(ActivityUserRepository $activityUserRepository)
+    {
+        $this->repository = $activityUserRepository;
+    }
+
+    public function checkRightsToActivity(Activity $activity, User $user): bool
```
Comment (SpartakusMd) [inline src/Security/AccessRightsPolicy.php]: Rename to `canAccessActivity`

**Review by SpartakusMd — CHANGES_REQUESTED** [review]

**Pull request merged**